30 setup wiremock

amitjangra2599 opened the pull request:
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.

Fixes #30 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Database change

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.

- [ ] Unit Tests
- [x] Integration Tests
- [ ] Manual Testing (please describe)

## Checklist:

- [x] I have created ADR (Architecture Decision Record) for this change, in case of architecture or tech stack changes.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes in other branches have been merged
- [x] I have updated the relevant application properties files (if needed)
- [ ] I have updated any necessary database migrations (if needed)

## Screenshots (for UX changes): NA

## Additional Notes:

Please add any additional notes, dependencies, or context that reviewers should know about.

SahilSaroha01 committed:
Wiremock config added for local and dev email test

amitjangra2599 committed:
Wiremock setup

amitjangra2599 committed:
Merge branch 'develop' of https://github.com/StatusNeo/vms into 30-setup-wiremock

# Conflicts:
#	web-backend/src/main/java/com/statusneo/vms/service/GraphEmailService.java
#	web-backend/src/main/java/com/statusneo/vms/service/OtpService.java
#	web-backend/src/main/resources/application.yml

amitjangra2599 committed:
Wiremock fix

amitjangra2599 committed:
Completed Wiremock setup

amitjangra2599 committed:
removed jetty dependency

amitjangra2599 committed:
Merge branch 'develop' of https://github.com/StatusNeo/vms into 30-setup-wiremock

amitjangra2599 committed:
externalized wiremock email url

amitjangra2599 committed:
Fix WireMock base URL resolution issue in integration tests

amitjangra2599 commented:
Hi @axymthr Please review this PR.

axymthr commented on web-backend/src/test/java/com/statusneo/vms/service/WiremockEmailServiceTest.java:
Try using the Wiremock JUnit Jupiter support instead http://wiremock.org/docs/junit-jupiter/

amitjangra2599 replied:
update WireMock tests to use JUnit Jupiter

axymthr reviewed (changes requested):
Can you add the Wiremock ADR to this PR? Enhance to clearly explain both why and how  we are using Wiremock in our setup, out of the various options available.

axymthr commented on web-backend/pom.xml:
```diff
@@ -152,6 +152,14 @@
+
+        <dependency>
+            <groupId>org.wiremock</groupId>
+            <artifactId>wiremock-standalone</artifactId>
```
Use plain `wiremock` dependency I believe this is the fat JAR? Let Spring BOM manage the version (if it can) for proper compatibility.

amitjangra2599 replied:
I initially switched to the plain wiremock-jre8 dependency, but it caused a java.lang.IncompatibleClassChangeError due to Jetty version conflicts with Spring Boot. Since wiremock-standalone isolates its dependencies (including Jetty), it avoids those issues in our test environment.

For now, I’ll continue using wiremock-standalone scoped to test to keep things stable. Open to revisiting this if we decide to align Jetty versions explicitly across the project later.

And this dependency is not managed by spring boot BOM, we need to manage its version manually.

axymthr replied:
Did you check out the official integration support? Any reason that wasn't working out? That seems to handle all this
https://github.com/wiremock/wiremock-spring-boot
https://www.wiremock.io/post/wiremock-now-has-an-official-spring-boot-integration
https://github.com/wiremock/wiremock-spring-boot

amitjangra2599 replied:
It worked thanks for the suggestion

amitjangra2599 committed:
test: migrate WireMock tests to JUnit Jupiter with @RegisterExtension

amitjangra2599 committed:
test: migrate to official WireMock Spring Boot integration with dynamic port support

- Replaced manual WireMock setup with @WireMockTest
- Used wiremock.server.port for dynamic URL resolution via application-test.yml
- Removed System.setProperty and @DynamicPropertySource

amitjangra2599 commented:
> Can you add the Wiremock ADR to this PR? Enhance to clearly explain both why and how we are using Wiremock in our setup, out of the various options available.

[ADR -Use WireMock.txt](https://github.com/user-attachments/files/21292903/ADR.-Use.WireMock.txt)

axymthr commented on web-backend/pom.xml:
```diff
@@ -152,6 +152,14 @@
+        <dependency>
+            <groupId>org.wiremock.integrations</groupId>
+            <artifactId>wiremock-spring-boot</artifactId>
+            <version>3.0.1</version>
```
See if it works without this dependency version

amitjangra2599 replied:
No, it's not working even without specifying the version. I also checked https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-dependencies/3.4.5, and the WireMock dependency is not managed by the Spring Boot BOM.

amitjangra2599 replied:
Hi @axymthr any update on this PR.

axymthr commented:
@amitjangra2599 the ADR doc needs to go in the folder https://github.com/StatusNeo/vms/tree/develop/web-backend/docs/adr in .md format. We have a gap in numbering so name it `0006-external-api-mock.md`

amitjangra2599 committed:
add architecture decision record for WireMock integration in tests

amitjangra2599 commented:
> @amitjangra2599 the ADR doc needs to go in the folder https://github.com/StatusNeo/vms/tree/develop/web-backend/docs/adr in .md format. We have a gap in numbering so name it `0006-external-api-mock.md`

Added

axymthr commented on web-backend/src/main/resources/application.yml:
```diff
@@ -41,4 +41,7 @@ gg:
+  admin-email: test@email.com
+
+mail:
+  url: http://localhost:${wiremock.server.port}/v1.0/users/%s/sendMail
```
This signals a problem with our Wiremock setup, basically its not properly reusable. Wiremock supports the concept of defining stubs in resources so they can be reused across the application. This URL is the same as the MS Graph URL, there has to be a cleaner way of doing this.

axymthr replied:
For now let's go ahead with this.

axymthr commented on web-backend/src/test/java/com/statusneo/vms/config/WireMockConfig.java:
```diff
@@ -0,0 +1,48 @@
+        wireMockServer.start();
+
+        logger.info("WireMock server started at port {}", wireMockServer.port());
+        System.setProperty("wiremock.mail.url", "http://localhost:" + wireMockServer.port()+"/v1.0/users/%s/sendMail");
```
This really feels like the wrong way to go about it.

amitjangra2599 replied:
WireMockConfig.java is outdated and has already been deleted

axymthr commented on web-backend/src/test/java/com/statusneo/vms/config/WireMockConfig.java:
```diff
@@ -0,0 +1,48 @@
+
+    }
+
+    @PreDestroy
```
I feel like this is not needed. Take a look at this reference test which is doing the same thing this code is doing, but seems like much less code.
https://github.com/wiremock/wiremock-spring-boot/blob/main/src/test/java/usecases/JavaStubbingTest.java

amitjangra2599 replied:
WireMockConfig.java is outdated and has already been deleted.
